README.md
---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->



# rarable

<!-- badges: start -->
[![R build status](https://github.com/basf/rarable/workflows/R-CMD-check/badge.svg)](https://github.com/basf/rarable/actions)
[![CRAN status](https://www.r-pkg.org/badges/version/rarable)](https://CRAN.R-project.org/package=rarable)
<!-- badges: end -->

Just an R client for [arable API](https://api.arable.cloud/api/v2/doc)
Provides access and functions to parse results into nice data.frames.

## Installation

You can install the development version of `rarable` using the [`remotes`](https://remotes.r-lib.org) package.


```r
remotes::install_github("basf/rarable")
```


## Usage

### Create client with basic authentication

Here we read the credentials from environmental variables.


```r
library("rarable")
ar <- arable$new(username = Sys.getenv("AR_USERNAME"),
  password = Sys.getenv("AR_PASSWORD"))
ar
```

But you can also use `apikey` or `token` authentication.

### List all devices & parse to a data.frame


```r
ar %>%
  ar_devices() %>%
  ar_parse_devices() %>%
  head()
```
### List all available tables


```r
ar %>%
  ar_schemas()
```

### List all columns of a table


```r
ar %>%
  ar_schema(table = "daily") %>%
  head()
```


### Retrieve data & parse to a data.frame


```r
ar %>%
  ar_data(
    table = "daily",
    device = "C007538",
    select = c("time", "mean_rh", "meant")) %>%
  ar_parse_data() %>%
  head()
```

## Unit tests

The package comes also with a suit of tests & coverage is >75%.
However, in order to run unit tests, you need to provide credentials as
environmental variables `AR_USERNAME`and `AR_PASSWORD`.


NEWS.md
# rarable 0.1.0

* First release


tests/testthat.R
library(testthat)
library(rarable)

test_check("rarable")


tests/testthat/test-data.R
if (nchar(Sys.getenv("AR_USERNAME")) == 0 ||
    nchar(Sys.getenv("AR_PASSWORD")) == 0)
  skip("No credentials for testing")

ar <- arable$new(username = Sys.getenv("AR_USERNAME"),
                 password = Sys.getenv("AR_PASSWORD"))

d <- ar_data(ar,
             table = "daily",
             limit = 50L,
             device = "C007538",
             select = c("time", "mean_rh"),
             start_time = Sys.time() - 60*60*24*30*12,
             end_time = Sys.time())

testthat::test_that("ar_data works", {
  expect_s3_class(d, "ar_data")
  expect_true(length(d[[1]]) == 2)
  expect_equal(names(d[[1]]), c("time", "mean_rh"))
})

dp <- ar_parse_data(d)
testthat::test_that("ar_parse_data works", {
  expect_s3_class(dp, "data.frame")
  expect_equal(names(dp), c("time", "mean_rh"))
})

## TODO local_time contains local_time column
d2 <- ar_data(ar,
             table = "daily",
             limit = 50L,
             device = "C007538",
             select = c("time", "mean_rh"),
             start_time = Sys.time() - 60*60*24*30*12,
             end_time = Sys.time(),
             local_time = "America/Regina"
             ) %>% ar_parse_data()

testthat::test_that("ar_data works with local_time", {
  expect_true("local_time" %in% names(d2))
})


tests/testthat/test-arable.R
testthat::test_that("arabvle fails with missing credentials", {
  expect_error(arable$new())
  arable$new(apikey = "")
})

if (nchar(Sys.getenv("AR_USERNAME")) == 0 ||
    nchar(Sys.getenv("AR_PASSWORD")) == 0)
  skip("No credentials for testing")


ar <- arable$new(username = Sys.getenv("AR_USERNAME"),
                 password = Sys.getenv("AR_PASSWORD"))

testthat::test_that("arable works", {
  expect_s3_class(ar, "R6")
  expect_s3_class(ar, "arable")
  expect_equal(ar$base_url, "https://api.arable.cloud/api/v2")
  expect_s3_class(ar$get_auth(), "request")
  expect_true("Authorization" %in% names(ar$get_auth()$headers))
  expect_true(ar$ping())
})

if (nchar(Sys.getenv("AR_APIKEY")) == 0)
  skip("No credentials for testing")

apk <- arable$new(apikey = Sys.getenv("AR_APIKEY"))

testthat::test_that("arable works with APIKEY", {
  expect_s3_class(apk, "R6")
  expect_s3_class(apk, "arable")
  expect_equal(apk$base_url, "https://api.arable.cloud/api/v2")
  expect_s3_class(apk$get_auth(), "request")
  expect_true("Authorization" %in% names(apk$get_auth()$headers))
  expect_true(apk$ping())
})


R/client.R
#' @title arable API client
#' @description handles the authentication scheme and provides a basic ping method.
#'
#' @export
#' @return An object of class `arable`.
#' @examples
#' \dontrun{
#' ar <- arable$new(username = Sys.getenv("AR_USERNAME"),
#'   password = Sys.getenv("AR_PASSWORD"))
#' ar
#' ar$get_auth()
#' ar$ping()
#'
#' arable$new(apikey = Sys.getenv("AR_APIKEY"))
#' }
arable <- R6::R6Class("arable",
  public = list(
    #' @field base_url root url to api
    base_url = "https://api.arable.cloud/api/v2",

    #' @description print method for `client`
    #' @param x self
    #' @param ... ignored
    print = function(...) {
      cat("  url: ", self$base_url, "\n", sep = "")

      con <- ifelse(is.null(private$auth$headers),
                    FALSE,
                    paste0(substr(private$auth$headers, 0, 15), "..."))

      cat("  Authentication:  ", con, "\n", sep = "")

      invisible(self)
    },


    #' @description Create a new client object using different methods.
    #' The Arable API provides three methods of authentication to the API:
    #' username + password (basic auth), apikey or bearer token.
    #' Precedance is token > apikey > username + password
    #' @param username username
    #' @param password password
    #' @param apikey apikey
    #' @param token token
    initialize = function(username,
        password,
        apikey,
        token) {
      if (!missing(token)) {
        private$token_auth(token)
        return(self)
      }
      if (!missing(apikey)) {
        private$apikey_auth(apikey)
        return(self)
      }
      if (!missing(username) & !missing(password)) {
        private$basic_auth(username, password)
        return(self)
      }
      stop("No credentials supplied!")
    },

    #' @description Get authentication header
    #' @return a httr request object
    get_auth = function() private$auth,

    #' @description Ping API by calling /devices endpoint.
    #' Returns TRUE if we get a 200
    ping = function() {
      resp <- httr::GET(file.path(self$base_url, "devices"),
                                query = list(page = 1L, limit = 1),
                                private$auth)
      return(httr::status_code(resp) == 200)
    }
  ),

  private = list(
    auth = NULL,

    apikey_auth = function(apikey) {
      private$auth <- httr::add_headers(
        Authorization = paste0("Apikey ", apikey))
    },

    token_auth = function(apikey) {
      private$auth <- httr::add_headers(
        Authorization = paste0("Bearer ", token))
    },

    basic_auth = function(username, password) {
      cred <- paste0(username, ":", password) %>%
        jsonlite::base64_enc()
      private$auth <- httr::add_headers(Authorization = paste0("Basic ", cred))
    }
  )
)


R/devices.R
#' List all devices
#' @param ar A \code{\link{arable}} object. Required.
#' @param page integer; starting page of api to receive data
#' @param limit integer; number of records per page (default: 24, max 10000)
#' @param verbose logical; verbose messages on API calls
#' @rdname devices
#' @export
#' @examples
#' \dontrun{
#' ar <- arable$new(username = Sys.getenv("AR_USERNAME"),
#'   password = Sys.getenv("AR_PASSWORD"))
#' devices <- ar_devices(ar, limit = 100L)
#' ar_parse_devices(devices)
#' }
ar_devices <- function(ar, page = 1L, limit = 24L, verbose = FALSE) {
  stopifnot(!missing(ar))
  stopifnot(inherits(ar, "arable"))
  stopifnot(is.integer(page))
  stopifnot(is.integer(limit))

  qurl <- file.path(ar$base_url, "devices")

  if (verbose) {
    message("GET: ", httr::modify_url(qurl,
                                      query = list(page = page, limit = limit)))
  }

  resp <- httr::GET(qurl,
                    query = list(page = page, limit = limit),
                    ar$get_auth())

  if (httr::http_type(resp) != "application/json") {
    stop("API did not return json", call. = FALSE)
  }

  parsed <- jsonlite::fromJSON(
    httr::content(resp, "text", encoding = "UTF-8"),
    simplifyVector = FALSE)

  if (httr::status_code(resp) != 200) {
    stop(
      sprintf(
        "Arable API request failed [%s]\n%s",
        httr::status_code(resp),
        parsed$message
      ),
      call. = FALSE
    )
  }

  out <- parsed$items
  next_page <- ifelse(parsed$page < parsed$pages,
                      parsed$page + 1,
                      NA)

  if (is.na(next_page)) {
    # base case (no next page)
    class(out) <- append("ar_devices", class(out))
    return(out)
  }

  # recursive case (next page available)
  out <- c(
    out,
    ar_devices(ar,
      page = as.integer(next_page),
      limit = limit,
      verbose = verbose)
  )

  class(out) <- append("ar_devices", class(out))
  return(out)
}

#' parse devices into a nice data.frame
#' @rdname devices
#' @param x a devices object, as returned by [ar_devices()]
#' @export
ar_parse_devices <- function(x) {
  stopifnot(inherits(x, "ar_devices"))
  purrr::map_df(x, parse_device)
}

parse_device <- function(device) {
  dev <- device[!vapply(device, is.list, FUN.VALUE = logical(1))]

  dev[is_empty(dev)] <- NA
  devd <- as.data.frame(dev)

  out <- data.frame(
    devd,
    location_id = device$current_location$id %|||% NA,
    location_name = device$current_location$name %|||% NA,
    location_longitude = device$current_location$gps[[1]] %|||% NA,
    location_latitude = device$current_location$gps[[2]] %|||% NA,
    location_altitude = device$current_location$elev %|||% NA,
    location_tzname = device$current_location$tz_name %|||% NA,
    location_start_date = device$current_location$start_date %|||% NA
  )[, c("type", "id", "org_id", "owner_id", "name", "firmware_id",
        "reported_fw", "last_seen", "last_post", "last_deploy", "state",
        "model", "sync_interval", "batt_pct", "batt_volt", "signal_strength",
        "has_bridge", "location_id", "location_name", "location_longitude",
        "location_latitude", "location_altitude", "location_tzname",
        "location_start_date")]
  return(out)
}


R/schemas.R
#' List all schemas
#' @param ar A \code{\link{arable}} object. Required.
#' @param verbose logical; verbose messages on API calls
#' @rdname schemas
#' @export
#' @examples
#' \dontrun{
#' ar <- arable$new(username = Sys.getenv("AR_USERNAME"),
#'   password = Sys.getenv("AR_PASSWORD"))
#' ar_schemas(ar)
#' }
ar_schemas <- function(ar, verbose = FALSE) {
  stopifnot(!missing(ar))
  stopifnot(inherits(ar, "arable"))

  qurl <- file.path(ar$base_url, "schemas")

  if (verbose) {
    message("GET: ", qurl)
  }

  resp <- httr::GET(qurl,
                    ar$get_auth())

  if (httr::http_type(resp) != "application/json") {
    stop("API did not return json", call. = FALSE)
  }

  parsed <- jsonlite::fromJSON(
    httr::content(resp, "text", encoding = "UTF-8"),
    simplifyVector = FALSE)

  if (httr::status_code(resp) != 200) {
    stop(
      sprintf(
        "Arable API request failed [%s]\n%s",
        httr::status_code(resp),
        parsed$message
      ),
      call. = FALSE
    )
  }

  return(unlist(parsed))
}

#' Show schema of a table
#' @param table tablename, as returned by [ar_schemas()]
#' @rdname schemas
#' @export
#' @examples
#' \dontrun{
#' ar <- arable$new(username = Sys.getenv("AR_USERNAME"),
#'   password = Sys.getenv("AR_PASSWORD"))
#' ar_schema(ar, "daily")
#' ar_schema(ar, "hourly")
#' }
ar_schema <- function(ar, table, verbose = FALSE) {
  stopifnot(!missing(ar))
  stopifnot(inherits(ar, "arable"))
  stopifnot(!missing(table))

  qurl <- file.path(ar$base_url, "schemas", table)

  if (verbose) {
    message("GET: ", qurl)
  }

  resp <- httr::GET(qurl,
                    ar$get_auth())

  if (httr::http_type(resp) != "application/json") {
    stop("API did not return json", call. = FALSE)
  }

  parsed <- jsonlite::fromJSON(
    httr::content(resp, "text", encoding = "UTF-8"),
    simplifyVector = FALSE)

  if (httr::status_code(resp) != 200) {
    stop(
      sprintf(
        "Arable API request failed [%s]\n%s",
        httr::status_code(resp),
        parsed$message
      ),
      call. = FALSE
    )
  }

  purrr::map_df(parsed, ~.x)
}


R/data.R
#' Get data
#' @param ar A \code{\link{arable}} object. Required.
#' @param table character; tablename, as returned by [ar_schemas()]. Required.
#' @param cursor cursor to start retrieval from
#' @param limit integer; limit; (1-10000)
#' @param order character; order
#' @param temp character; temp unit to use
#' @param pres chracter; pressure unit
#' @param ratio character; ratio unit
#' @param size character; size unit
#' @param speed character; speed unit
#' @param vol charcter; volume unit
#' @param volumetric_soil_moisture_unit character;
#' @param device character; device
#' @param location character; location
#' @param local_time character; local time
#' @param select character vector; columns to select, see [ar_schema()]
#'   for options of a table
#' @param start_time datetime in POSIXct class
#' @param end_time datetime in POSIXct class
#' @param verbose logical; verbose messages on API calls
#' @param recurse logical; should the function recurse until
#'   all data is fetched?
#' @rdname data
#' @export
#' @examples
#' \dontrun{
#' ar <- arable$new(username = Sys.getenv("AR_USERNAME"),
#'   password = Sys.getenv("AR_PASSWORD"))
#' ar_data(ar,
#'     table = "daily",
#'     limit = 50L,
#'     device = "C007538",
#'     select = c("time", "mean_rh"),
#'     start_time = Sys.time() - 60*60*24*30*12,
#'     end_time = Sys.time(),
#'     verbose = TRUE)
#' }
ar_data <- function(ar,
    table,
    cursor = NULL,
    limit = 1000L,
    order = c("asc", "desc"),
    temp = c("C", "F"),
    pres = c("mb", "kp"),
    ratio = c("dec", "pct"),
    size = c("in", "mm"),
    speed = c("mps", "kph", "mph"),
    vol = c("g", "l"),
    volumetric_soil_moisture_unit = c("millimeter_per_meter",
                                      "inches_per_foot"),
    device = NULL,
    location = NULL,
    local_time = NULL,
    select = NULL,
    start_time = NULL,
    end_time = NULL,
    verbose = FALSE,
    recurse = TRUE) {

  stopifnot(!missing(ar))
  stopifnot(inherits(ar, "arable"))
  stopifnot(!missing(table))
  stopifnot((!is.null(device) & is.null(location)) |
              (is.null(device) & !is.null(location)))
  stopifnot(is.integer(limit))
  stopifnot(1L <= limit & limit <= 10000L)

  start_time_f <- NULL
  if (!is.null(start_time)) {
    stopifnot(lubridate::is.POSIXct(start_time))
    start_time_f <- format(start_time, "%Y-%m-%dT%H:%M:%SZ")
  }
  end_time_f <- NULL
  if (!is.null(end_time)) {
    stopifnot(lubridate::is.POSIXct(end_time))
    end_time_f <- format(end_time, "%Y-%m-%dT%H:%M:%SZ")
  }

  order <- match.arg(order)
  temp <- match.arg(temp)
  pres <- match.arg(pres)
  ratio <- match.arg(ratio)
  size <- match.arg(size)
  speed <- match.arg(speed)
  vol <- match.arg(vol)
  volumetric_soil_moisture_unit <- match.arg(volumetric_soil_moisture_unit)

  if (!is.null(select))
    select <- paste(select, collapse = ",")


  qurl <- file.path(ar$base_url, "data", table)
  q <- list(
    cursor = cursor,
    limit = limit,
    order = order,
    temp = temp,
    pres = pres,
    ratio = ratio,
    size = size,
    speed = speed,
    vol = vol,
    volumetric_soil_moisture_unit = volumetric_soil_moisture_unit,
    device = device,
    location = location,
    select = select,
    start_time = start_time_f,
    end_time = end_time_f,
    local_time = local_time) %>%
    purrr::compact()

  if (verbose) {
    message("GET: ", httr::modify_url(qurl,
                                      query = q))
  }

  resp <- httr::GET(qurl,
                    query = q,
                    ar$get_auth())

  if (httr::http_type(resp) != "application/json") {
    stop("API did not return json", call. = FALSE)
  }

  parsed <- jsonlite::fromJSON(
    httr::content(resp, "text", encoding = "UTF-8"),
    simplifyVector = FALSE)

  if (httr::status_code(resp) != 200) {
    stop(
      sprintf(
        "Arable API request failed [%s]\n%s",
        httr::status_code(resp),
        parsed$message
      ),
      call. = FALSE
    )
  }

  out <- parsed

  if (recurse) {
    next_cursor <- resp$headers$`x-cursor-next`

    if (is.null(next_cursor)) {
      # base case (no next page)
      class(out) <- append("ar_data", class(out))
      return(out)
    }

    # recursive case (next page available)
    out <- c(
      out,
      ar_data(ar,
              table,
              cursor = next_cursor,
              limit = limit,
              order = order,
              temp = temp,
              pres = pres,
              ratio = ratio,
              size = size,
              speed = speed,
              vol = vol,
              volumetric_soil_moisture_unit = volumetric_soil_moisture_unit,
              device = device,
              location = location,
              local_time = local_time,
              select = select,
              start_time = start_time,
              end_time = end_time,
              verbose = verbose,
              recurse = recurse)
    )
  }

  class(out) <- append("ar_data", class(out))
  return(out)
}


#' parse data into a nice data.frame
#' @rdname data
#' @param x a data object, as returned by [ar_data()]
#' @export
ar_parse_data <- function(x) {
  stopifnot(inherits(x, "ar_data"))
  purrr::map_df(x, function(y) {
    y[is_empty(y)] <- NA
    data.frame(y)
    })
}
